### PR TITLE
Errors

### PR DESCRIPTION
**Pull request by mardy**
Creates Error classes from XML annotations.

Adds an optional parameter to specify the namespace where the error classes should be generated, in case the default one creates conflicts.

### CODE REVIEW

**Comment by jonte** [comment]
Why not use the annotation format mentioned in https://dbus.freedesktop.org/doc/dbus-specification.html#introspection-format ?

Seems like we could do something like this:
```
<method name="Frobate">
    <arg name="foo" type="i" direction="in"/>
    <arg name="bar" type="s" direction="out"/>
    <arg name="baz" type="a{us}" direction="out"/>
    <!-- Documentation of error -->
    <annotation name="org.gdbus.codegen.glibmm.Error" value="MyCustomError"/>
</method>
```
Might want to pick a more clever annotation name though..

File: codegen_glibmm/templates/common.cpp.templ [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    if (g_str_has_prefix(gerror->message, prefix)) {
+        message = strstr(gerror->message + sizeof(prefix), ": ") + 2;
+    } else {
+        message = gerror->message;
+    }
+
+    throw Error(gerror->code, message);
+}
+
+static Glib::Quark registerDomain()
```
Comment (jonte) [inline codegen_glibmm/templates/common.cpp.templ]: should this be `register_domain` ?

**Comment by mardy** [comment]
> Why not use the annotation format mentioned in https://dbus.freedesktop.org/doc/dbus-specification.html#introspection-format ?

I think that this is a very good idea. For some reason, other toolkits didn't go this way, but I think we should give it a try. Maybe like this:

    
    <annotation name="error" value="org.gdbus.codegen.glibmm.Error.MyCustomError">

Then I suggest that we mandate that this annotations should be made at the interface level, and not per method (you can still write the list of errors emitted by a method in a documentation comment): the reason is that we might find the same error mentioned in several methods, and the documentation line might be different, among these instances. It's a problem we have also in my solution, but I think that if we use the `annotation` keyword, then it's easier to solve it.

**Comment by jonte** [comment]
We could even use a different field than `value=` I guess, maybe `error-name=` to make it more clear that `name=` is the name of the annotation, and the second parameter is the name of the value.

**Comment by mardy** [comment]
> Why not use the annotation format mentioned in https://dbus.freedesktop.org/doc/dbus-specification.html#introspection-format ?

I think that this is a very good idea. For some reason, other toolkits didn't go this way, but I think we should give it a try. Maybe like this:

    
    <annotation name="error" value="org.gdbus.codegen.glibmm.Error.MyCustomError">

Then I suggest that we mandate that this annotations should be made at the interface level, and not per method (you can still write the list of errors emitted by a method in a documentation comment): the reason is that we might find the same error mentioned in several methods, and the documentation line might be different, among these instances. It's a problem we have also in my solution, but I think that if we use the `annotation` keyword, then it's easier to solve it.

> We could even use a different field than `value=` I guess, maybe `error-name=` to make it more clear that `name=` is the name of the annotation, and the second parameter is the name of the error.

Unfortunately we cannot do this, because the DTD for the D-Bus XML files only admits `name` and `value` as attributes for annotations.

File: codegen_glibmm/dbustypes.py [hunk 2]
```diff
@@ -34,6 +34,18 @@ def __init__(self, lines):
         self.lines = lines
 
 
+class Error:
+    def __init__(self, value):
+        (prefix, name) = value.rsplit('.', 1)
```
Comment (jonte) [inline codegen_glibmm/dbustypes.py]: What if the interface write choses a non-dotted error name? like (in the XML) `value="MyError"` ?

File: tests/data/simple/input_stub.cpp [hunk 3]
```diff
@@ -2,6 +2,18 @@ static const char interfaceXml0[] = R"XML_DELIMITER(<!DOCTYPE node PUBLIC "-//fr
                       "http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
 <node>
   <interface name="org.gdbus.codegen.glibmm.Test">
+    <!--
+      Permission to perform the action has been denied. But let's not
+      despair and continue writing this error description to make it span
```
Comment (jonte) [inline tests/data/simple/input_stub.cpp]: :D

**Comment by jonte** [comment]
Looks great, I think. Can you also make an entry to the README explaining how the errors work?

I guess we need to add something similar to this:

> --cpp-namespace=NAMESPACE
> A string to prepend to the namespaces of the generated functions.

And also maybe a note on the `<annotation` usage.

**Pull request merged**